lib/disposable.dart

```diff
@@ -179,63 +182,83 @@ class Disposable implements _Disposable, DisposableManager {
     }
     _isDisposing = true;
 
-    List<Future> futures = []
-      ..addAll(_internalDisposables.map((disposable) => disposable.dispose()))
-      ..add(onDispose());
-
-    _internalDisposables = [];
-
+    await Future.wait(_blockingFutures);
     // We need to filter out nulls because a subscription cancel
     // method is allowed to return a plain old null value.
-    return Future
-        .wait(futures.where((future) => future != null))
-        .then(_completeDisposeFuture);
+    await Future.wait(_internalDisposables
+        .map((disposable) => disposable.dispose())
+        .where((future) => future != null));
+
+    _internalDisposables = null;
+    _blockingFutures = null;
+
+    return onDispose().then(_completeDisposeFuture);
   }
 
   /// Automatically dispose another object when this object is disposed.
   ///
   /// The parameter may not be `null`.
   @mustCallSuper
   @override
-  void manageDisposable(Disposable disposable) {
-    _throwIfNull(disposable, 'disposable');
+  Disposable manageDisposable(Disposable disposable) {
+    _throwOnInvalidCall(disposable, 'disposable');
     _internalDisposables.add(disposable);
+    return disposable;
   }
 
   /// Automatically handle arbitrary disposals using a callback.
   ///
   /// The parameter may not be `null`.
   @mustCallSuper
   @override
-  void manageDisposer(Disposer disposer) {
-    _throwIfNull(disposer, 'disposer');
+  void manageDisposer<T>(Disposer disposer) {
+    _throwOnInvalidCall(disposer, 'disposer');
     _internalDisposables.add(new _InternalDisposable(disposer));
   }
 
+  /// Causes the object to add the future to a list of futures which will
+  /// be awaited before the object disposes itself or any of the objects
+  /// under its management.
+  @mustCallSuper
+  @override
+  Future<T> manageFuture<T>(Future<T> future) {
```

Comment:
Yeah, I had originally named it `awaitFutureBeforeDispose` or something like that. I'm open to a better name. I'm not sold on something as simple as `wait`, though, since I think that would suggest something more like `Future.wait`.